### ISSUE TITLE
Out of bounds read when parsing an RRCP packet capture

### ISSUE THREAD

**Issue by mgv** [the-tcpdump-group/tcpdump#581]
Hi there!

This bug was found with AFL and tested against https://github.com/the-tcpdump-group/tcpdump/tree/17a3c288589e18b9f1299fdf350852c768e2e887

Base64 encoded pcap file:
```
1MOyoTAwMDAwMDAwMDAwMGAwMABrAAAwMDAwMDAwMDBKAAAAMDAwMDAwRYiZsDAwMDAwMDAwMDAw
MDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAw
MDAwMDAwMDAIAAAAMDAwMAIIMDmIVTAwMA==
```

Run `tcpdump -r file.pcap` to reproduce the issue.

```
(gdb) run -r rrcp_crash 
Starting program: /home/vagrant/tcpdump_git/tcpdump/tcpdump -r rrcp_crash
reading from file rrcp_crash, link-type FRELAY (Frame Relay)

Program received signal SIGSEGV, Segmentation fault.
etheraddr_string (ndo=ndo@entry=0xbfffe458, ep=ep@entry=0xb792efff <error: Cannot access memory at address 0xb792efff>) at ./addrtoname.c:486
486             tp = lookup_emem(ndo, ep);
(gdb) bt
#0  etheraddr_string (ndo=ndo@entry=0xbfffe458, ep=ep@entry=0xb792efff <error: Cannot access memory at address 0xb792efff>) at ./addrtoname.c:486
#1  0x080a47b5 in rrcp_print (ndo=ndo@entry=0xbfffe458, cp=cp@entry=0xb792efff <error: Cannot access memory at address 0xb792efff>, length=length@entry=808464441) at ./print-rrcp.c:90
#2  0x0806cf4b in ethertype_print (ndo=ndo@entry=0xbfffe458, ether_type=ether_type@entry=34969, p=0xb792f00d "\260", '\060' <repeats 68 times>, length=808464427, caplen=808464427) at ./print-ether.c:372
#3  0x0807105d in fr_print (ndo=0xbfffe458, p=0xb792f008 "00E\210\231\260", '\060' <repeats 68 times>, length=808464432) at ./print-fr.c:276
#4  0x080511c3 in pretty_print_packet (ndo=ndo@entry=0xbfffe458, h=h@entry=0xbfffe330, sp=sp@entry=0xb792f008 "00E\210\231\260", '\060' <repeats 68 times>, packets_captured=1) at ./print.c:339
#5  0x0804d907 in print_packet (user=0xbfffe458 "", h=0xbfffe330, sp=0xb792f008 "00E\210\231\260", '\060' <repeats 68 times>) at ./tcpdump.c:2492
#6  0xb7e05b26 in ?? () from /usr/lib/i386-linux-gnu/libpcap.so.0.8
#7  0xb7df5e1f in pcap_loop () from /usr/lib/i386-linux-gnu/libpcap.so.0.8
#8  0x0804c4b8 in main (argc=3, argv=0xbffff694) at ./tcpdump.c:1996
```

The faulty assignment seems to take place on line 372 of the `print_ether.c` file.

```
case ETHERTYPE_RRCP:
        rrcp_print(ndo, p - 14 , length + 14);
	return (1);
```

`p - 14` ends up pointing outside of the mapped regions, causing a SIGSEGV on the next read.

**Comment by mgv** [comment]
I'm having some issues reproducing this on a fresh libpcap/tcpdump build, so I might have linked an older libpcap before. I'll close the issue and reopen with an updated crash file if I am able to reproduce it.

**Comment by guyharris** [comment]
Denis, I think one of your changes fixed this.

**Comment by guyharris** [comment]
> I'm having some issues reproducing this on a fresh libpcap/tcpdump build, so I might have linked an older libpcap before. I'll close the issue and reopen with an updated crash file if I am able to reproduce it.

The fuzzing affected the file header in ways that older libpcaps didn't catch and new libpcaps do catch.  You'd have to #if 0-out the code in `sf-pcap.c` that catches those problems to try to test the file with newer versions of libpcap.

**Comment by guyharris** [comment]
Denis's fix is in the security branch (Will be merged in a few weeks).

**Comment by infrastation** [comment]
The original reporter for this vulnerability is Hanno Böck. It took a long time to publish the fix so the same issue was later independently discovered by other researchers.

Commit 6bc4429 fixes this and the fix is available in the release 4.9.0. Thanks to everybody for their reports, please send any future security reports to security@tcpdump.org.

**Comment by guyharris** [comment]
Duplicate of #560
